homeassistant/components/modbus/validators.py

```diff
@@ -40,97 +42,90 @@
 
 _LOGGER = logging.getLogger(__name__)
 
-ENTRY = namedtuple("ENTRY", ["struct_id", "register_count"])
+ENTRY = namedtuple(
+    "ENTRY",
+    [
+        "struct_id",
+        "register_count",
+        "validate_parm",
+    ],
+)
+CONV_PARM = {
+    CONF_COUNT: 0,
+    CONF_STRUCTURE: 1,
+    CONF_SLAVE_COUNT: 2,
+    CONF_SWAP_BYTE: 3,
+    CONF_SWAP_WORD: 4,
+    CONF_SWAP_WORD_BYTE: 4,
+}
 DEFAULT_STRUCT_FORMAT = {
-    DataType.INT8: ENTRY("b", 1),
-    DataType.INT16: ENTRY("h", 1),
-    DataType.INT32: ENTRY("i", 2),
-    DataType.INT64: ENTRY("q", 4),
-    DataType.UINT8: ENTRY("c", 1),
-    DataType.UINT16: ENTRY("H", 1),
-    DataType.UINT32: ENTRY("I", 2),
-    DataType.UINT64: ENTRY("Q", 4),
-    DataType.FLOAT16: ENTRY("e", 1),
-    DataType.FLOAT32: ENTRY("f", 2),
-    DataType.FLOAT64: ENTRY("d", 4),
-    DataType.STRING: ENTRY("s", 1),
+    DataType.INT8: ENTRY("b", 1, [False, False, False, False, False]),
+    DataType.UINT8: ENTRY("c", 1, [False, False, False, False, False]),
+    DataType.INT16: ENTRY("h", 1, [False, False, True, True, False]),
+    DataType.UINT16: ENTRY("H", 1, [False, False, True, True, False]),
+    DataType.FLOAT16: ENTRY("e", 1, [False, False, True, True, False]),
+    DataType.INT32: ENTRY("i", 2, [False, False, True, True, True]),
+    DataType.UINT32: ENTRY("I", 2, [False, False, True, True, True]),
+    DataType.FLOAT32: ENTRY("f", 2, [False, False, True, True, True]),
+    DataType.INT64: ENTRY("q", 4, [False, False, True, True, True]),
+    DataType.UINT64: ENTRY("Q", 4, [False, False, True, True, True]),
+    DataType.FLOAT64: ENTRY("d", 4, [False, False, True, True, True]),
+    DataType.STRING: ENTRY("s", 1, [True, False, False, False, False]),
+    DataType.CUSTOM: ENTRY("?", 0, [True, True, False, False, False]),
```

Comment:
This one is a bit hard to following.  I'm not sure what `validate_parm` means here.  Maybe make it a NamedTuple as well?

Reply:
Validate_parms checks 5 keywords which are only legal with some datatypes.

e.g. "swap: word" is not allowed for "datatype: int16", I will add a comment above to explain that. 

I could make it a NamedTuple just thought is was a lot of overhead, but will do.

Reply:
namedtuple is very cheap unless you create 10000s of them